feat: move integration tests to main repo

axiomatic-aardvark opened the pull request:
Adds e2e tests to the main repo. 

To run all the tests, use the provided shell script `./run-e2e-tests.sh`.
To run individual tests, first spin up the relevant setup instances, for instance for most test it's enough to have a few basic instances of the POI Radio running:
```
docker-compose -f e2e-tests.docker-compose.yml up basic-instance  --scale basic-instance=5
```

Then, run the desired test:
```
cargo run --bin integration-tests -- --check=poi_ok
```

`poi_ok` (probably needs a better name) is the basic but most important sanity check test, it spins up another normal Radio instance that waits until it receives 5 unique messages, then exits successfully.

Other tests like `invalid_payload` and `poi_divergence_local` require a different setup, all of that is demonstrated in the shell script and can be used as reference.

I've added the shell script as a CI check so that if any of our e2e test fails (or times out) we get an error. Moving forward, we should add a new e2e test for every **bigger** new feature. It's important to keep in mind that the tests instrument Docker images so are quite heavy and slow, that's why we should be careful about adding too much, they're main point is to make sure that the whole system works, and not so much to check individual bits of logic. Anything that can be just a unit test - should be just a unit test. 

Currently the tests take about 30 minutes to run when starting from scratch and 13-15 minutes when you already have the Docker images built.

**Next steps**
Right now the tests are just a Rust binary crate, but we should convert them to actual cargo tests with cfg(test), that will be a bit tricky but is totally doable. That will enable us to fully sandbox the e2e tests and make changes to both the SDK and the Radio that would only be compiled if they're being ran in a test environment.

Also, the current `test-radio.rs` is just a slightly edited version of `poi-radio`'s main.rs file. In the interest of time I'm keeping it that way but we should refactor `main.rs` to be more modular so that `test-radio.rs` can just reuse the things it needs. That way updating `test-radio.rs` after the POI Radio's main logic is changed won't be needed.

hopeyen commented on integration-tests/src/checks/correct_filtering_different_topics.rs:
```diff
@@ -0,0 +1,66 @@
+            "No message found with topic {}",
+            MOCK_SUBGRAPH_MAINNET
+        );
+        assert!(
```
Trying to think if this should be a bit more strict, maybe we could switch to if there are no messages other than the configured subgraphs, instead of only checking for MOCK_SUBGRAPH_GOERLI?

axiomatic-aardvark replied:
✅

hopeyen commented on integration-tests/src/checks/invalid_block_hash.rs:
```diff
@@ -0,0 +1,48 @@
+    if messages.len() > 0 {
+        error!("{}", "invalid_block_hash test failed".red());
+        std::process::exit(1);
+    }
```
do we want a success log printed and exit with code 0?

axiomatic-aardvark replied:
some of the validation one (I think it was this one, invalid_nonce and invalid_sender) are specifically checking for messages to **not** be in the store after a certain interval of listening (currently 1m), so really these 3 tests are treated as successful if they time out, not if they exit with 0

hopeyen commented on integration-tests/src/checks/num_messages.rs:
```diff
@@ -0,0 +1,55 @@
+        "Expected message arr length to be at least 70% of mock senders count."
+    );
+
+    info!("{}", "num_messages test is sucessful ✅".green());
```
succccessful ✨  (There are a couple more but saving the comments here as you could ctrl+ to replace `sucessful`

axiomatic-aardvark replied:
✅

hopeyen commented on integration-tests/src/checks/poi_divergence_local.rs:
```diff
@@ -0,0 +1,89 @@
+                    info!("Most attested npoi: {:#?}", most_attested_npoi);
+                    info!("Local npoi: {:#?}", &local_attestation.npoi);
+
+                    // This check assumes there are less basic than divergent instances running. This is what remote attestations looks like currently with no divergence (example):
```
do we want this comment? could the attestation details be tested with expected stake sum and sender addr?

axiomatic-aardvark replied:
ideally they would, but since the network can lag behind the messages we get are not deterministic, that's why I left the check a bit loose, but you're right the comment just adds noise, will remove

axiomatic-aardvark replied:
✅

hopeyen commented on integration-tests/src/checks/store_attestations.rs:
```diff
@@ -0,0 +1,91 @@
+    remote: &RemoteAttestationsMap,
+    local: &LocalAttestationsMap,
+) {
+    debug!("Something from test_attestation_handler");
```
Is this debug! used for something?

axiomatic-aardvark replied:
not anymore 😅  will remove

axiomatic-aardvark replied:
✅

hopeyen commented on integration-tests/src/utils.rs:
```diff
@@ -0,0 +1,222 @@
+    );
+}
+
+pub struct RadioRuntimeConfig {
```
Do you think these could maybe be renamed? The folder structure is clear that these config is for the tests, but I'm worried as it sounds like config for normal runtime. Maybe something like...`RadioTestConfig`?

axiomatic-aardvark replied:
agreed

axiomatic-aardvark replied:
also docstrings would be helpful I suppose

axiomatic-aardvark replied:
✅

hopeyen reviewed (commented):
Great work, this is a lot of stuff!! 

I made a brief pass and just have some minor comments

axiomatic-aardvark commented on .github/workflows/main.yml:
```diff
@@ -107,4 +105,38 @@ jobs:
+    if: github.event.review.state == 'approved'
+    runs-on: ubuntu-latest
+    env:
+      GRAPH_NODE_STATUS_ENDPOINT: "http://localhost:8030/graphql"
```
the environment variable is set because it's required, but it's not being used later on, all graph node responses are mocked

axiomatic-aardvark commented on src/graphql/mod.rs:
```diff
@@ -2,7 +2,13 @@ use graphql_client::{GraphQLQuery, Response};
+use graphcast_sdk::{graphql::QueryError, BlockPointer};
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct SubgraphStatus {
```
✅

axiomatic-aardvark commented on integration-tests/src/setup/test_radio.rs:
```diff
@@ -0,0 +1,500 @@
+    // Parse basic configurations
+    let config = Config::args();
+
+    let collect_message_duration: i64 = env::var("COLLECT_MESSAGE_DURATION")
```
yes, but I don't want to run Config::args in `test.radio.rs` since it's already called in `integration-tests/main.rs`. It is easier and makes more sense to use it this way imo because the setup instances set `COLLECT_MESSAGE_DURATION` as an env var too.

axiomatic-aardvark replied:
I think if we make the config editable globally in the future that would help with switching all internal env vars to it, but it's not pressing

axiomatic-aardvark commented on integration-tests/src/setup/test_radio.rs:
```diff
@@ -0,0 +1,500 @@
+        &network_subgraph,
+        &graph_node_endpoint,
+        read_boot_node_addresses(config.boot_node_addresses.clone()),
+        Some("5"),
```
updated ✅  though we should probably switch to an enum for that soon

axiomatic-aardvark commented on integration-tests/src/checks/clear_store.rs:
```diff
@@ -0,0 +1,47 @@
+            "There were messages found with block {block} and subgraph id {subgraph}"
+        );
+
+        assert!(messages.lock().unwrap().to_vec().len() < prev_len);
```
when we get to this assertion, a comparison is guaranteed to have happened, otherwise the test would've timed out (considered as a failure)

axiomatic-aardvark commented on integration-tests/src/checks/clear_store.rs:
```diff
@@ -0,0 +1,47 @@
+
+        assert!(messages.lock().unwrap().to_vec().len() < prev_len);
+
+        info!("{}", "clear_store test is sucessful ✅".green());
```
✅

axiomatic-aardvark commented on integration-tests/src/checks/comparison_interval.rs:
```diff
@@ -0,0 +1,57 @@
+    let elapsed = start_time.elapsed();
+    let one_minute = Duration::from_secs(60);
+
+    if elapsed >= one_minute {
```
we are using collect message duration 🤔  but yeah I suppose this can be a more specific test

axiomatic-aardvark replied:
✅

axiomatic-aardvark commented:
@hopeyen  thanks for the awesome feedback! 🚀  I implemented most of the changes and left some comments on things that I'm not sure if we should change (at least right now). Also added this issue - https://github.com/graphops/poi-radio/issues/64 , where we can add further improvement ideas.

axiomatic-aardvark commented on integration-tests/src/setup/test_radio.rs:
```diff
@@ -0,0 +1,500 @@
+            .expect("Could not register handler");
+    };
+
+    let mut block_store: HashMap<NetworkName, BlockClock> = HashMap::new();
```
✅

axiomatic-aardvark commented on integration-tests/src/setup/test_radio.rs:
```diff
@@ -0,0 +1,500 @@
+        env::var("GRAPH_NODE_STATUS_ENDPOINT").expect("No Graph node status endpoint provided.");
+
+    // Send message every x blocks for which wait y blocks before attestations
+    let wait_block_duration = 2;
```
✅
